evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -158,43 +166,67 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None) -> STLDecomposer:
             self.logger.warning(
                 f"STLDecomposer may perform poorly on data with a high seasonal smoother ({self.seasonal_smoother}).",
             )
+        X, y = self._check_target(X, y)
 
-        # Save the frequency of the fitted series for checking against transform data.
-        self.frequency = y.index.freqstr or pd.infer_freq(y.index)
-
-        # Determine the period of the seasonal component
-        if self.period is None:
-            self.set_period(X, y)
-
-        stl = STL(y, seasonal=self.seasonal_smoother, period=self.period)
-        res = stl.fit()
-        self.seasonal = res.seasonal
-        self.period = stl.period
-        dist = len(y) % self.period
-        self.seasonality = (
-            self.seasonal[-(dist + self.period) : -dist]
-            if dist > 0
-            else self.seasonal[-self.period :]
-        )
-        self.trend = res.trend
-        self.residual = res.resid
+        if isinstance(y, pd.Series):
+            y = y.to_frame()
+
+        # If there is a series_id in stacked data or more than one column in unstacked data, set multiseries to true
+        is_multiseries = False
+        if self.series_id or len(y.columns) > 1:
+            is_multiseries = True
+
+        # Iterate through each id group
+        self.decompositions = {}
+        for id in y.columns:
+            series_y = y[id]
+            self.original_index = series_y.index if series_y is not None else None
+
+            X, series_y = self._check_target(X, series_y)
+
+            self._map_dt_to_integer(self.original_index, series_y.index)
+
+            # Save the frequency of the fitted series for checking against transform data.
+            self.frequency = series_y.index.freqstr or pd.infer_freq(series_y.index)
+            # Determine the period of the seasonal component
+            if is_multiseries or self.period is None:
+                self.set_period(X, series_y)
+
+            stl = STL(series_y, seasonal=self.seasonal_smoother, period=self.period)
+            res = stl.fit()
+            self.seasonal = res.seasonal
+            self.period = stl.period
+            dist = len(series_y) % self.period
+            self.seasonality = (
+                self.seasonal[-(dist + self.period) : -dist]
+                if dist > 0
+                else self.seasonal[-self.period :]
+            )
+
+            self.trend = res.trend
+            self.residual = res.resid
+
+            if is_multiseries:
+                self.decompositions[id] = {
+                    "seasonal": self.seasonal,
+                    "seasonality": self.seasonality,
+                    "trend": self.trend,
+                    "residual": self.residual,
+                    "period": self.period,
+                }
 
         return self
 
     def transform(
         self,
         X: pd.DataFrame,
-        y: pd.Series = None,
-    ) -> tuple[pd.DataFrame, pd.Series]:
+        y: pd.DataFrame = None,
+    ):
```

Comment:
y can be a Series or a DataFrame, and it looks like the return types got deleted here - they're still useful!